### PR TITLE
feat: implements dynamic configurations pattern and their validation logic

### PR DESCRIPTION
**Pull request by baranbartu**
- Implements dynamic dag-like question mechanism for docker compose and helm
- Writes a schema to restrict the scope of the questions and to validate properly
- Implements validation mechanism against schema
- Implements validation mechanism based on whether or not there is a dead-end in dag-like questions
- Implements CI pipeline for PR checks - validate questions

### CODE REVIEW

**Review by Younday — COMMENTED** [review]
A few small changes here and there, the PR looks good in general with the dynamic configuration

File: dynamic-configuration/src/validation.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import sys
+
+import yaml
+from jsonschema import Draft7Validator
+
+schema = yaml.safe_load(open("./src/questions-schema.yaml", "r"))
```
[the remark this answers is not in the document]
Comment (Younday) [reply dynamic-configuration/src/validation.py:6]: Let me rephrase this a bit, the schema should probably stay together with the validation script, but we can possibly accept the actual question files from other paths.

File: dynamic-configuration/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+packages = [{include = "src", from = "./"}]
+
+[tool.poetry.dependencies]
+python = "^3.12"
+jsonschema = "^4.22.0"
+pyyaml = "^6.0.1"
+
+[tool.poetry.group.dev.dependencies]
+ipdb = "^0.13.13"
+ipython = "^8.23.0"
```
[the remark this answers is not in the document]
Comment (baranbartu) [reply dynamic-configuration/pyproject.toml:20]: actually I use it all the time to quickly to test something locally.
```
poetry shell
ipython
--> test some code quickly
```

File: dynamic-configuration/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+flake8 = "^7.0.0"
```
[the remark this answers is not in the document]
Comment (baranbartu) [reply dynamic-configuration/pyproject.toml:21]: I need to also adjust my dev ecosystem. I'll find its vim plugin and add it. But until then it is really helping me when developing on my vim ecosystem. Can I keep it until I have ruff in my dev env as well? It is for dev dependency already.

**Comment by Younday** [comment]
Maybe another addition on the approach: we can use something like [datamodel_code_generator](https://docs.pydantic.dev/latest/integrations/datamodel_code_generator/ ) for the schema YAML file to create the objects in Pydantic and export those as Python classes via this module.  We can then use those classes in other scripts to validate the given configuration for example

File: dynamic-configuration/pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Younday) [reply dynamic-configuration/pyproject.toml:21]: Yeah they're dev dependencies anyways, so it's fine to keep for now

File: dynamic-configuration/src/validation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (baranbartu) [reply dynamic-configuration/src/validation.py:6]: do you have a sample skeleton in mind?

[the remark this answers is not in the document]
Comment (baranbartu) [reply dynamic-configuration/src/validation.py:11]: would you prefer this separation now? And if so, how would you envision the structure?

**Comment by Younday** [comment]
> > Maybe another addition on the approach: we can use something like [datamodel_code_generator](https://docs.pydantic.dev/latest/integrations/datamodel_code_generator/) for the schema YAML file to create the objects in Pydantic and export those as Python classes via this module. We can then use those classes in other scripts to validate the given configuration for example
> 
> What would be an example use case for this approach?

That would be mostly to have a generated client/definitions based on the JSON file. In this case we can export that, and use it in other parts of the application. It is maybe a bit overkill with the Pydantic classes exposed as well, but that would make the usage of this in other parts (or other repo's) a bit easier. I did not do my full change here, I did want to add an example as well of what that will look like. 

Do you see any upsides to this approach over your approach or not?

**Review by younes-syntho — APPROVED** [review]
LGTM
